sdk/src/marginCalculation.ts

```diff
@@ -0,0 +1,283 @@
+		this.strict = false;
+		this.ignoreInvalidDepositOracles = false;
+		this.marginBuffer = new Map();
+	}
+
+	static standard(marginType: MarginCategory): MarginContext {
+		return new MarginContext(marginType);
+	}
+
+	static liquidation(marginBuffer: Map<number | 'cross', BN>): MarginContext {
+		const ctx = new MarginContext('Maintenance');
+		ctx.mode = { type: 'Liquidation' };
+		ctx.marginBuffer = marginBuffer;
+		return ctx;
+	}
+
+	strictMode(strict: boolean): this {
+		this.strict = strict;
+		return this;
+	}
+
+	ignoreInvalidDeposits(ignore: boolean): this {
+		this.ignoreInvalidDepositOracles = ignore;
+		return this;
+	}
+
+	setMarginBuffer(buffer: Map<number | 'cross', BN>): this {
+		this.marginBuffer = buffer;
+		return this;
+	}
+}
+
+export class IsolatedMarginCalculation {
+	marginRequirement: BN;
+	totalCollateral: BN; // deposit + pnl
+	totalCollateralBuffer: BN;
+	marginRequirementPlusBuffer: BN;
+
+	constructor() {
+		this.marginRequirement = new BN(0);
+		this.totalCollateral = new BN(0);
+		this.totalCollateralBuffer = new BN(0);
+		this.marginRequirementPlusBuffer = new BN(0);
+	}
+
+	getTotalCollateralPlusBuffer(): BN {
+		return this.totalCollateral.add(this.totalCollateralBuffer);
+	}
+
+	meetsMarginRequirement(): boolean {
+		return this.totalCollateral.gte(this.marginRequirement);
+	}
+
+	meetsMarginRequirementWithBuffer(): boolean {
+		return this.getTotalCollateralPlusBuffer().gte(
+			this.marginRequirementPlusBuffer
+		);
+	}
+
+	marginShortage(): BN {
+		const shortage = this.marginRequirementPlusBuffer.sub(
+			this.getTotalCollateralPlusBuffer()
+		);
+		return shortage.isNeg() ? new BN(0) : shortage;
+	}
+}
+
+export class MarginCalculation {
+	context: MarginContext;
+	totalCollateral: BN;
+	totalCollateralBuffer: BN;
+	marginRequirement: BN;
+	marginRequirementPlusBuffer: BN;
+	isolatedMarginCalculations: Map<number, IsolatedMarginCalculation>;
+	allDepositOraclesValid: boolean;
+	allLiabilityOraclesValid: boolean;
+	withPerpIsolatedLiability: boolean;
+	withSpotIsolatedLiability: boolean;
+	totalSpotLiabilityValue: BN;
+	totalPerpLiabilityValue: BN;
+	trackedMarketMarginRequirement: BN;
+	fuelDeposits: number;
+	fuelBorrows: number;
+	fuelPositions: number;
+
+	constructor(context: MarginContext) {
+		this.context = context;
+		this.totalCollateral = new BN(0);
+		this.totalCollateralBuffer = new BN(0);
+		this.marginRequirement = new BN(0);
+		this.marginRequirementPlusBuffer = new BN(0);
+		this.isolatedMarginCalculations = new Map();
+		this.allDepositOraclesValid = true;
+		this.allLiabilityOraclesValid = true;
+		this.withPerpIsolatedLiability = false;
+		this.withSpotIsolatedLiability = false;
+		this.totalSpotLiabilityValue = new BN(0);
+		this.totalPerpLiabilityValue = new BN(0);
+		this.trackedMarketMarginRequirement = new BN(0);
+		this.fuelDeposits = 0;
+		this.fuelBorrows = 0;
+		this.fuelPositions = 0;
+	}
+
+	addCrossMarginTotalCollateral(delta: BN): void {
+		const crossMarginBuffer =
```

Comment:
nit: might wanna just store the crossMarginBuffer on the context so you dont have to do this `??` every time?